### ISSUE TITLE
[Performance] Linear Layer Benchmark

### ISSUE THREAD

**Issue by cccyf** [ml-explore/mlx#1276]
**Describe the question**
Hi, I benchmarked the Linear layer with MLX on M1 Max, and compared that with Pytorch on H100. The trend seems different. I wonder if MLX Linear has the potential to perform more stably with an increase in sequence length?

**Benchmark Code to Reproduce**
**MLX**
```python
import sys

import mlx.core as mx
import mlx.nn
import time

def benchmark_linear_projection(
    input_dims: int, output_dims: int, q_len: int, dtype: mx.Dtype
) -> None:
    proj = mlx.nn.Linear(input_dims, output_dims, bias=False)
    proj.set_dtype(dtype)
    x = mx.random.uniform(shape=(1, q_len, input_dims)).astype(dtype)
    mx.eval(proj, x)
    # warm up
    for _ in range(10):
        mx.eval(proj(x))

    iteration = 32 
    tic = time.perf_counter()
    def time_linear(proj, x):
        for _ in range(iteration):
            mx.eval(proj(x))

    time_linear(proj, x)
    timing = 1e3 * (time.perf_counter() - tic)
    print(f"run mlx linear takes {timing / iteration: .3f} (ms)")

if __name__ == "__main__":
    mx.random.seed(123)
    for dtype in (mx.float16, mx.bfloat16):
        with open(f"/tmp/linear_{str(dtype).split('.')[-1]}.log", "w") as sys.stdout:
            for seq_len in range(1, 101, 1):
                print(f"benchmark seq_len {seq_len}")
                benchmark_linear_projection(4096, 4096, seq_len, dtype)
```
**Pytorch**
```python
import sys
import time

import torch
from recurrent_drafting import attention, rng

def benchmark_linear_projection(
    input_dims: int, output_dims: int, q_len: int, dtype: torch.dtype
) -> None:
    device = torch.device("cuda")
    proj = torch.nn.Linear(input_dims, output_dims, bias=False).to(dtype).to(device)
    x = torch.randn(size=(1, q_len, input_dims)).to(dtype).to(device)
    # warm up
    for _ in range(10):
        proj(x)
    
    torch.cuda.synchronize(device)
    tic = time.perf_counter()
    iteration = 32
    for _ in range(iteration):
        proj(x)
    
    torch.cuda.synchronize(device)
    timing = 1e3 * (time.perf_counter() - tic)
    print(f"run torch linear takes {timing/iteration: .3f} (ms)")


if __name__ == "__main__":
    torch.manual_seed(123)
    torch.cuda.manual_seed(123)
    for dtype in (torch.float16, torch.bfloat16):
        with open(f"/tmp/linear_{str(dtype).split('.')[-1]}.log", "w") as sys.stdout:
            for seq_len in range(5, 102, 5):
                print(f"benchmark seq_len {seq_len}")
                benchmark_linear_projection(4096, 4096, seq_len, dtype)
```

**Benchmark Result**
![benchmark_linear](https://github.com/user-attachments/assets/478911a9-bbbf-4ff9-bcb4-b4bc621c7c31)
Logs:
1. MLX
[linear_bfloat16_mlx.log](https://github.com/user-attachments/files/16315050/linear_bfloat16_mlx.log)
[linear_float16_mlx.log](https://github.com/user-attachments/files/16315052/linear_float16_mlx.log)

**Expected behavior**
Latency of MLX Linear is gradually increasing with longer sequence length. However, latency of pytorch Linear on H100 seems pretty stable. 

**Desktop (please complete the following information):**
 - OS Version: MacOS 14.5
 - MLX Version 
```
mlx                       0.16.0          py310h481d4a5_0    conda-forge
```

**Comment by angeloskath** [comment]
Hi! There are a few things to uncover here.

1. Latency should increase for the H100 as well. However, the "sequence length" is still small so it doesn't reach the linear scaling part yet. It has to do with the specifics of the kernel and how GPUs actually run. For instance you 'll see that MLX takes a bit more time every 32 rows. That has to do with the tile size of the matmul in this particular case which means that whether we run `(24, 4096) x (4096, 4096)` or `(31, 4096) x (4096, 4096)` the gpu will run exactly the same code and it will take the same amount of time.
2. It won't make too big of a difference but running eval every iteration in the loop is equal to running `torch.cuda.synchronize` inside every iteration. My MacBook Air goes from 2.08 TFLOPS to 2.68 for a sequence length of 64. I changed `proj(x)` to `x = proj(x)` and then one `mx.eval(x)`.

To sum up, the time should increase linearly with the size of the input matrix. Due to the way GPUs work it will increase linear but with steps like the behavior you observe in MLX. The width and height of these steps will be different for the H100 which is a much much bigger GPU than a laptop's.

Running for sizes `[i*128 for i in range(1, 129)]` will probably give you the same linear increase on the H100.

I will close the issue but feel free to reopen if you think I didn't explain the observed behavior satisfactorily.

**Comment by cccyf** [comment]
Thanks @angeloskath for the explanation! That makes a lot of sense to me.

Meanwhile, I'm wondering if the latency increase of Linear layer in the chart looks as expected? If I understand correctly, with every additional 32 rows in the input, the latency of one linear call increases by 0.2 ms?
